tests/test_options.py

```diff
@@ -1455,31 +1688,252 @@ def cli(one, two):
         (
-            False,  # multiple
-            2,  # nargs
-            42,  # default (not a list)
-            "'default' must be a list when 'nargs' != 1.",
+            {"flag_value": Class1, "default": True},
+            {"flag_value": Class2},
+            ["--opt1"],
+            "<class 'test_options.Class1'>",
         ),
         (
-            True,  # multiple
-            2,  # nargs
-            [
-                "test string which is not a list in the list"
-            ],  # default (list of non-lists)
-            "'default' must be a list of lists when 'multiple' is true"
-            " and 'nargs' != 1.",
+            {"flag_value": Class1, "default": True},
+            {"flag_value": Class2},
+            ["--opt2"],
+            "<class 'test_options.Class2'>",
+        ),
+        # Even the default is processed as a string.
+        ({"flag_value": Class1, "default": "True"}, {"flag_value": Class2}, [], "True"),
+        ({"flag_value": Class1, "default": None}, {"flag_value": Class2}, [], None),
+        # To get the classes as-is, we need to specify the type as UNPROCESSED.
+        (
+            {"flag_value": Class1, "type": UNPROCESSED, "default": True},
+            {"flag_value": Class2, "type": UNPROCESSED},
+            [],
+            True,
+        ),
+        (
+            {"flag_value": Class1, "type": UNPROCESSED, "default": True},
+            {"flag_value": Class2, "type": UNPROCESSED},
+            ["--opt1"],
+            Class1,
+        ),
+        (
+            {"flag_value": Class1, "type": UNPROCESSED, "default": True},
+            {"flag_value": Class2, "type": UNPROCESSED},
+            ["--opt2"],
+            Class2,
+        ),
+        # Setting the default to a class, an instance of the class is returned instead
+        # of the class itself, because the default is allowed to be callable (and
+        # consummd). And this what happens the type is.
+        (
+            {"flag_value": Class1, "default": Class1},
+            {"flag_value": Class2},
+            [],
+            re.compile(r"'<test_options.Class1 object at 0x[0-9A-Fa-f]+>'"),
+        ),
+        (
+            {"flag_value": Class1, "default": Class2},
+            {"flag_value": Class2},
+            [],
+            re.compile(r"'<test_options.Class2 object at 0x[0-9A-Fa-f]+>'"),
+        ),
+        (
+            {"flag_value": Class1, "type": UNPROCESSED, "default": Class1},
+            {"flag_value": Class2, "type": UNPROCESSED},
+            [],
+            re.compile(r"<test_options.Class1 object at 0x[0-9A-Fa-f]+>"),
+        ),
+        (
+            {"flag_value": Class1, "type": UNPROCESSED, "default": Class2},
+            {"flag_value": Class2, "type": UNPROCESSED},
+            [],
+            re.compile(r"<test_options.Class2 object at 0x[0-9A-Fa-f]+>"),
+        ),
+        ###
+        ### Reproduces https://github.com/pallets/click/issues/2012#issuecomment-892437060
```

Comment:
Describe what's being tested, not only a link.

Reply:
Done in: https://github.com/pallets/click/pull/3030/commits/7b84990ead4a663e4b57d5cdb83eca1e508c3d78